lib/pkcs11_req.c

```diff
@@ -1151,20 +1166,23 @@ int pkcs11_sign_X509_REQ(pkcs11Context * p11Context, CK_VOID_PTR req, int output
 	    goto err;
 	}
 
-	/* just do right housekeeping */
-	if(sig->r !=NULL) {  BN_free(sig->r); sig->r = NULL; }
-	if(sig->s !=NULL) {  BN_free(sig->s); sig->s = NULL; }
-
+	if((sig_r = BN_new()) == NULL) {
+      goto err;
+  }
+  if((sig_s = BN_new()) == NULL) {
+      goto err;
+  }
 
-	if( (sig->r = BN_bin2bn( &outbuf[0], outlen/2, sig->r)) == NULL) {
+	if( BN_bin2bn( &outbuf[0], outlen/2, sig_r) == NULL) {
 	    P_ERR();
 	    goto err;
 	}
 
-	if( (sig->s = BN_bin2bn( &outbuf[outlen/2], outlen/2, sig->s)) == NULL ) {
+	if( BN_bin2bn( &outbuf[outlen/2], outlen/2, sig_s) == NULL ) {
 	    P_ERR();
 	    goto err;
 	}
+	DSA_SIG_set0(sig, sig_r, sig_s);
```

Comment:
`sig_r`and `sig_s` must be set to `NULL` after this call, to avoid double free condition (once cleanup code for these variables is added after the `err:`label, see previous comments)